### DISCUSSION TITLE
curl response to proxy authentication 407

### DISCUSSION THREAD

**Discussion by huaraz** [curl/curl#15718] [General]
Looking at the details https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/407 it says:

_This response is sent with a [Proxy-Authenticate](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Proxy-Authenticate) header that contains information on how to correctly authenticate requests. The client may repeat the request with a new or replaced [Proxy-Authorization](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Proxy-Authorization) header field._

When I use curl against a http URL I get error 0 for a 407 response, whereas with a https URL I get error 56 and a 407 response.  I think both should be treated the same as  a valid response from the proxy i.e. not an error.  That the body is empty in case of https is not really relevant.

Any thoughts on this ?

Thank you
Markus

**Comment by pkropachev** [comment]
Usually, to access HTTPS endpoint `CONNECT` method is used to tunnel any kind of connection through an HTTP proxy.

Tunnel is not created in case attempt access HTTP endpoint via proxy. So, you just receive response code 407 due to incorrect authentification on proxy.

```
curl -x http://auth-proxy.testdomain.com:3128 http://httpbin.testdomain.com/get -v -k -o /dev/null
* Host auth-proxy.testdomain.com:3128 was resolved.
* IPv6: (none)
* IPv4: 192.168.222.4
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 192.168.222.4:3128...
* Connected to auth-proxy.testdomain.com (192.168.222.4) port 3128
* using HTTP/1.x
> GET http://httpbin.testdomain.com/get HTTP/1.1
> Host: httpbin.testdomain.com
> User-Agent: curl/8.11.1-DEV
> Accept: */*
> Proxy-Connection: Keep-Alive
> 
* Request completely sent off
< HTTP/1.1 407 Proxy Authentication Required
< Server: squid/6.6
< Mime-Version: 1.0
< Date: Wed, 18 Dec 2024 10:44:39 GMT
< Content-Type: text/html;charset=utf-8
< Content-Length: 3573
< X-Squid-Error: ERR_CACHE_ACCESS_DENIED 0
< Vary: Accept-Language
< Content-Language: en
< Proxy-Authenticate: Basic realm="Squid proxy-caching web server"
< Cache-Status: 921cf4bc7900
< Connection: keep-alive
< Via: squid/6.1
< 
{ [3573 bytes data]
100  3573  100  3573    0     0  1970k      0 --:--:-- --:--:-- --:--:-- 3489k
* Connection #0 to host auth-proxy.testdomain.com left intact
```

In case sending request to HTTPS endpoint via proxy, curl creates HTTP tunnel using `CONNECT` method, but establishing tunnel is failed.

```
curl -x http://auth-proxy.testdomain.com:3128 https://httpbin.testdomain.com/get -v -k -o /dev/null
* Host auth-proxy.testdomain.com:3128 was resolved.
* IPv6: (none)
* IPv4: 192.168.222.4
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 192.168.222.4:3128...
* CONNECT tunnel: HTTP/1.1 negotiated
* allocate connect buffer
* Establish HTTP proxy tunnel to httpbin.testdomain.com:443
> CONNECT httpbin.testdomain.com:443 HTTP/1.1
> Host: httpbin.testdomain.com:443
> User-Agent: curl/8.11.1-DEV
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 407 Proxy Authentication Required
< Server: squid/6.6
< Mime-Version: 1.0
< Date: Wed, 18 Dec 2024 10:44:13 GMT
< Content-Type: text/html;charset=utf-8
< Content-Length: 3551
< X-Squid-Error: ERR_CACHE_ACCESS_DENIED 0
< Vary: Accept-Language
< Content-Language: en
< Proxy-Authenticate: Basic realm="Squid proxy-caching web server"
< Cache-Status: 921cf4bc7900
< Connection: keep-alive
< Via: squid/6.1
< 
* Ignore 3551 bytes of response-body
* CONNECT tunnel failed, response 407
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* closing connection #0
curl: (56) CONNECT tunnel failed, response 407
```

So, cases are a bit different. What do you observe in your case?

**Reply by huaraz** [reply]
But I think it is a wrong behaviour. The client must not start the tunnel until it received a 200 code.  Any other valid code returned is not an error.  There are only 2 error conditions 

1) No valid http status code is returned
2) 200 is returned but tunnel failed e.g. TLS handshake error (if TLS is used over connect which is not a must like when tunneling ssh)

The client must negotiate the authentication with the proxy and prompt the user for username / password , not error out and provide the same response as for http

* Connection #0 to host auth-proxy.testdomain.com left intact

Markus

**Reply by icing** [reply]
Trying to parse this. In the first example by @pkropachev there is no tunnel created. Curl uses the proxy to perform the request, which gives a 407 response. The second example tries to create a tunnel, which fails with a 407, and the request is never sent.

So, the first gives a response to the request which is 407.
The second fails to establish the connection, does not send the request and thus, gets no response to the request.

Does that highlight the differences?

**Reply by huaraz** [reply]
I think a tunnel creation must not be attempted until a 200 response is received from the proxy.

**Reply by icing** [reply]
Thanks, in our tests, it does not, neither for http: or https: resources. Could you give a curl command line example where you think it happens? And are you using the current curl or an older version?

**Reply by huaraz** [reply]
The example from @pkropachev shows it i.e. gives error 56. 

curl: (56) CONNECT tunnel failed, response 407

I think it should return no error leave it open like for http 

* Connection #0 to host auth-proxy.testdomain.com left intact

If I use libcurl with curl_easy_perform with a https URL I get an error (like with curl )  and can't extract the  status code and header to perform the correct authentication handshake.

**Reply by icing** [reply]
I think if the tunnel could not be established, this should be an error. libcurl cannot show a HTTP status code for the transfer, since the 407 is not the response to the request.

Can you not inspect the headers via https://curl.se/libcurl/c/curl_easy_header.html ?

**Reply by bagder** [reply]
It causes an error because you ask to do a transfer that fails. Failing to setup the tunnel through the proxy means the transfer is not performed => failure.

When you do a GET to a proxy or server and that returns a response, that is a *successful* transfer containing a 4xx code. A different thing.

**Reply by huaraz** [reply]
I still disagree. A empty body response with a correct response header code 407 should still be valid for a CONNECT method. I tried the curl_easy_header but after an error it seems empty I even don't know if after an unsuccessful  curl_easy_perform to a https URL I can call curl_easy_header. Can you show an example code where you can negotiate the proxy authentication for a https connection ?  The proxy give you a choice of NTLM or Negotiate authentication and you need to e.g. chose NTLM. I am struggling doing this as simple as possible.

Thank you

**Reply by pkropachev** [reply]
> I tried the curl_easy_header but after an error it seems empty I even don't know if after an unsuccessful curl_easy_perform to a https URL I can call curl_easy_header.

You can.

```cpp
#include <curl/curl.h>

int main() {
  curl_global_init(CURL_GLOBAL_DEFAULT);
  CURL *curl;
  curl = curl_easy_init();
  FILE *output = fopen("/dev/null", "w");
  struct curl_header *header;

  if (curl) {
    curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
    curl_easy_setopt(curl, CURLOPT_URL, "https://example.com");
    curl_easy_setopt(curl, CURLOPT_PROXY, "http://auth-proxy.testdomain.com:8080");
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, output);

    auto res = curl_easy_perform(curl);

    if (res != CURLE_OK) {
      fprintf(stdout, "curl_easy_perform() failed: %d, %s\n", res,
              curl_easy_strerror(res));
    }

    fprintf(stdout, "----------------------------\n");

    if (curl_easy_header(curl, "Proxy-Authenticate", 0, CURLH_CONNECT, -1,
                         &header) == CURLHE_OK) {
      fprintf(stdout, "Proxy-Authenticate: %s\n", header->value);

      if (header->amount == 2 &&
          curl_easy_header(curl, "Proxy-Authenticate", 1, CURLH_CONNECT, -1,
                           &header) == CURLHE_OK) {
        fprintf(stdout, "Proxy-Authenticate: %s\n", header->value);
      }
    }

    fprintf(stdout, "----------------------------\n");

    fprintf(stdout, "All server headers:\n");
    {
      struct curl_header *h;
      struct curl_header *prev = NULL;
      do {
        h = curl_easy_nextheader(curl, CURLH_CONNECT, -1, prev);
        if (h)
          fprintf(stdout, " %s: %s (%u)\n", h->name, h->value, (int)h->amount);
        prev = h;
      } while (h);
    }
    curl_easy_cleanup(curl);
  }

  return 0;
}
```

Output:

```
./curl-test
* Host auth-proxy.testdomain.com:8080 was resolved.
* IPv6: (none)
* IPv4: 172.18.187.146
*   Trying 172.18.187.146:8080...
* CONNECT tunnel: HTTP/1.1 negotiated
* allocate connect buffer
* Establish HTTP proxy tunnel to example.com:443
> CONNECT example.com:443 HTTP/1.1
Host: example.com:443
Proxy-Connection: Keep-Alive

< HTTP/1.1 407 Proxy Authentication Required
< Content-Type: text/html
< Content-Length: 58062
< Connection: keep-alive
< Proxy-Authenticate: NTLM
< Proxy-Authenticate: Basic realm="netentsec"
< 
* Ignore 58062 bytes of response-body
* CONNECT tunnel failed, response 407
* closing connection #0
curl_easy_perform() failed: 56, Failure when receiving data from the peer
----------------------------
Proxy-Authenticate: NTLM
Proxy-Authenticate: Basic realm="netentsec"
----------------------------
All server headers:
 Content-Type: text/html (1)
 Content-Length: 58062 (1)
 Connection: keep-alive (1)
 Proxy-Authenticate: NTLM (2)
 Proxy-Authenticate: Basic realm="netentsec" (2)
```

In this case `curl_easy_header()` returns only one `Proxy-Authenticate` header, but proxy provides two. So, I use `curl_easy_nextheader()`. Also I guess some proxy servers can return single `Proxy-Authenticate` and set supported types of authentification via comma (e.g. `Proxy-Authenticate: NTLM, Basic...`). You can check values and add corresponding curl_easy option to authentificate on proxy.

**UPD**: Extract both values from `Proxy-Authenticate` using `curl_easy_header()`.

**Reply by bagder** [reply]
> In this case curl_easy_header() returns only one Proxy-Authenticate header, but proxy provides two

You need to check the struct it fills in with data about the returned header. The `amount` field should be 2 if there are two headers using the same name, and then you can decide which of the instances you want to ask for with the `index` argument to the `curl_easy_header()`  function.

Or use `curl_easy_nextheader` to just iterate over them all.

**Reply by huaraz** [reply]
Oh I see the curl_easy_header function does not use CURL_HEADER which I expected it to be. What is the reason to have a CURLH_CONNECT ?  To me both are response headers.

Thank you

**Reply by huaraz** [reply]
Is there also a special option to get the connect response code with curl_easy_getinfo ? CURL_INFO_RESPONSE_CODE has 0 for a connect instead of 407.

**Reply by bagder** [reply]
[CURLINFO_HTTP_CONNECTCODE](https://curl.se/libcurl/c/CURLINFO_HTTP_CONNECTCODE.html)

They have to be different since you often get both.

**Reply by bagder** [reply]
> What is the reason to have a CURLH_CONNECT ? To me both are response headers.

The reason is that you need to tell curl *which* headers you want. You might get the same header name in both sets for example.

**Reply by huaraz** [reply]
I see. Thank you !
